File: generateTrees95.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        
        return helper(1, n);
    }
    
    private List<TreeNode> helper(int min, int max) {
        List<TreeNode> result = new ArrayList<TreeNode>();
        if (min > max) return result;
        
        for(int rt = min; rt <= max; rt++) {
            List<TreeNode> leftList = helper(min, rt-1);
            List<TreeNode> rightList = helper(rt+1, max);
            if (leftList.size() == 0 & rightList.size() == 0) {
                TreeNode root = new TreeNode(rt);
                result.add(root);
            }
            else if (leftList.size() == 0) {
                for (TreeNode right : rightList) {
                    TreeNode root = new TreeNode(rt);
                    root.right = right;
                    result.add(root);
                }
            }
            else if (rightList.size() == 0) {
                for (TreeNode left : leftList) {
                    TreeNode root = new TreeNode(rt);
                    root.left = left;
                    result.add(root);
                }
            }
            else {
                for (TreeNode left : leftList) {
                    for (TreeNode right : rightList) {
                        TreeNode root = new TreeNode(rt);
                        root.left = left;
                        root.right = right;
                        result.add(root);
                    }
                }
            }
        }
        return result;
    }
}

File: generateTrees95.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def helper(self, min, max):
        if min > max: return [None]
        
        res = []
        
        for curr in range(min, max + 1):
            left = self.helper(min, curr - 1)
            right = self.helper(curr + 1, max)
            
            for l in left:
                for r in right:
                    root = TreeNode(curr)
                    root.left = l
                    root.right = r
                    res.append(root)
                    
        return res
        
    def generateTrees(self, n):
        
        if n == 0:
            return []
        else:
            return self.helper(1, n)


print(Solution().generateTrees(3))